### ISSUE TITLE
logger=ngalert error="runtime error: invalid memory address or nil pointer dereference"

### ISSUE THREAD

**Issue by jan666** [grafana/grafana#47093]
**What happened**:

I'm using the FreeBSD grafana8 port (https://www.freshports.org/www/grafana8/). It was recently updated to 8.3.6_1 and now builds with Go 1.18. After I re-build my grafana 8 it crashes on start

```
t=2022-03-31T07:38:16+0200 lvl=warn msg="ngalert feature flag is deprecated: use unified alerting enabled setting instead" logger=settings
t=2022-03-31T07:38:16+0200 lvl=warn msg="falling back to legacy setting of 'min_interval_seconds'; please use the configuration option in the `unified_alerting` section if Grafana 8 alerts are enabled." logger=settings
t=2022-03-31T07:38:16+0200 lvl=info msg="Config loaded from" logger=settings file=/usr/local/share/grafana/conf/defaults.ini
t=2022-03-31T07:38:16+0200 lvl=info msg="Config loaded from" logger=settings file=/usr/local/etc/grafana/grafana.ini
t=2022-03-31T07:38:16+0200 lvl=info msg="Path Home" logger=settings path=/usr/local/share/grafana
t=2022-03-31T07:38:16+0200 lvl=info msg="Path Data" logger=settings path=/var/db/grafana
t=2022-03-31T07:38:16+0200 lvl=info msg="Path Logs" logger=settings path=/var/log/grafana
t=2022-03-31T07:38:16+0200 lvl=info msg="Path Plugins" logger=settings path=/var/db/grafana/plugins
t=2022-03-31T07:38:16+0200 lvl=info msg="Path Provisioning" logger=settings path=/usr/local/etc/grafana/provisioning
t=2022-03-31T07:38:16+0200 lvl=info msg="App mode production" logger=settings
t=2022-03-31T07:38:16+0200 lvl=info msg="Connecting to DB" logger=sqlstore dbtype=sqlite3
t=2022-03-31T07:38:16+0200 lvl=warn msg="SQLite database file has broader permissions than it should" logger=sqlstore path=/var/db/grafana/grafana.db mode=-rw-r--r-- expected=-rw-r-----
t=2022-03-31T07:38:16+0200 lvl=info msg="Starting DB migrations" logger=migrator
t=2022-03-31T07:38:16+0200 lvl=info msg="migrations completed" logger=migrator performed=0 skipped=377 duration=985.679µs
t=2022-03-31T07:38:16+0200 lvl=info msg="Initialising plugins" logger=plugin.manager
t=2022-03-31T07:38:16+0200 lvl=info msg="Plugin registered" logger=plugin.manager pluginId=input
t=2022-03-31T07:38:16+0200 lvl=info msg="Plugin registered" logger=plugin.manager pluginId=grafana-clock-panel
t=2022-03-31T07:38:16+0200 lvl=info msg="Plugin registered" logger=plugin.manager pluginId=natel-discrete-panel
t=2022-03-31T07:38:16+0200 lvl=info msg="Live Push Gateway initialization" logger=live.push_http
t=2022-03-31T07:38:16+0200 lvl=info msg="warming cache for startup" logger=ngalert
t=2022-03-31T07:38:16+0200 lvl=info msg="HTTP Server Listen" logger=http.server address=192.168.42.254:4000 protocol=http subUrl= socket=
t=2022-03-31T07:38:16+0200 lvl=eror msg="rule not found for instance, ignoring" logger=ngalert rule=BHoshvHnk
t=2022-03-31T07:38:16+0200 lvl=info msg="starting MultiOrg Alertmanager" logger=ngalert.multiorg.alertmanager
t=2022-03-31T07:38:26+0200 lvl=eror msg="alert rule panic" logger=ngalert error="runtime error: invalid memory address or nil pointer dereference" stack="goroutine 341 [running]:\nruntime/debug.Stack()\n\truntime/debug/stack.go:24 +0x65\ngithub.com/grafana/grafana/pkg/services/ngalert/eval.executeQueriesAndExpressions.func1()\n\tgithub.com/grafana/grafana/pkg/services/ngalert/eval/eval.go:259 +0x5d\npanic({0x5c9f40, 0x4819ae0})\n\truntime/panic.go:838 +0x207\nreflect.mapiterinit(0x95e7c0?, 0xc000db71c0?, 0xc000db71c0?)\n\truntime/map.go:1373 +0x19\ngonum.org/v1/gonum/graph/iterator.(*mapIter).next(0xc000e3d6a0)\n\tgonum.org/v1/gonum/graph/iterator/map.go:89 +0x5e\ngonum.org/v1/gonum/graph/iterator.(*Nodes).Next(0xc000d9b740)\n\tgonum.org/v1/gonum/graph/iterator/nodes_map.go:40 +0x36\ngithub.com/grafana/grafana/pkg/expr.buildNodeRegistry(0xc0016419e0)\n\tgithub.com/grafana/grafana/pkg/expr/graph.go:115 +0x149\ngithub.com/grafana/grafana/pkg/expr.(*Service).buildDependencyGraph(0xc000db7288?, 0x26e34b7?)\n\tgithub.com/grafana/grafana/pkg/expr/graph.go:85 +0x28\ngithub.com/grafana/grafana/pkg/expr.(*Service).buildPipeline(0xc0896d749b797f94?, 0x26a4b126c?)\n\tgithub.com/grafana/grafana/pkg/expr/graph.go:65 +0x19\ngithub.com/grafana/grafana/pkg/expr.(*Service).BuildPipeline(...)\n\tgithub.com/grafana/grafana/pkg/expr/service.go:61\ngithub.com/grafana/grafana/pkg/expr.(*Service).TransformData(0xc0015426c0, {0x108d3f0, 0xc000d88ea0}, 0xc000d9aa80)\n\tgithub.com/grafana/grafana/pkg/expr/transform.go:134 +0x114\ngithub.com/grafana/grafana/pkg/services/ngalert/eval.executeQueriesAndExpressions({0x1, 0x1, {0x82c35d480, 0xc001526d60}, {0x108d3f0, 0xc000d88ea0}}, {0xc0015e6f00?, 0x4abe160?, 0x5348e0?}, {0xc0896d749b797f94, ...}, ...)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/eval/eval.go:274 +0x1ce\ngithub.com/grafana/grafana/pkg/services/ngalert/eval.executeCondition({0x1, 0x1, {0x82c35d480, 0xc001526d60}, {0x108d3f0, 0xc000d88ea0}}, 0xc000db7c18, {0xc0896d749b797f94, 0x26a4b126c, 0x4abe160}, ...)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/eval/eval.go:173 +0xed\ngithub.com/grafana/grafana/pkg/services/ngalert/eval.(*Evaluator).ConditionEval(0xc0010e57f8, 0xc000db7c18, {0x29ef6f2?, 0xc0?, 0x4abe160?}, 0xd0?)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/eval/eval.go:510 +0x16f\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine.func2(0xc00025cc30, 0xacd84f?, 0xc000aa25c0)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:460 +0x30a\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine.func4.2(0x38?)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:556 +0x1b7\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine.func3(0xc0012e3c00)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:517 +0x5c\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine.func4(0x0?, 0x300000002?, {0xc000b94901?, {0xc000c4e5a0?, 0x26bc59b?}}, 0xc000aa25c0, 0xc000d3deb0, 0xc001628c30, 0xc0012e3bc0, {0x109b0b0, ...}, ...)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:546 +0x194\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine(0xc0010e57a0, {0x108d380, 0xc000e790c0}, {0xc0010e57a8?, {0xc000c4e5a0?, 0xc000c5df88?}}, 0xc000c47bc0)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:561 +0x6b5\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).schedulePeriodic.func1()\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:367 +0x38\ngolang.org/x/sync/errgroup.(*Group).Go.func1()\n\tgolang.org/x/sync/errgroup/errgroup.go:57 +0x67\ncreated by golang.org/x/sync/errgroup.(*Group).Go\n\tgolang.org/x/sync/errgroup/errgroup.go:54 +0x8d\n"
t=2022-03-31T07:38:26+0200 lvl=info msg="level=debug component=dispatcher msg=\"Received alert\" alert=DatasourceError[5406912][active]" logger=alertmanager org=1
```

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Grafana version: 8.3.6
- Data source type & version: InfluxDB 1.8
- OS Grafana is installed on: FreeBSD 13
- User OS & Browser: macOS / Safari
- Grafana plugins:
- Others:

**Comment by fliiiix** [comment]
I can reproduce this if I enable the [new alerts](https://grafana.com/docs/grafana/latest/alerting/unified-alerting/opt-in/) on my instance see

```
t=2022-03-31T20:16:03+0200 lvl=eror msg="alert rule panic" logger=ngalert error="runtime error: invalid memory address or nil pointer dereference" stack="goroutine 749 [running]:\nruntime/debug.Stack()\n\truntime/debug/stack.go:24 +0x65\ngithub.com/grafana/grafana/pkg/services/ngalert/eval.executeQueriesAndExpressions.func1()\n\tgithub.com/grafana/grafana/pkg/services/ngalert/eval/eval.go:259 +0x5d\npanic({0x5c9f40, 0x4819ae0})\n\truntime/panic.go:838 +0x207\nreflect.mapiterinit(0x95e7c0?, 0xc00191f1c0?, 0xc00191f1c0?)\n\truntime/map.go:1373 +0x19\ngonum.org/v1/gonum/graph/iterator.(*mapIter).next(0xc00171efb0)\n\tgonum.org/v1/gonum/graph/iterator/map.go:89 +0x5e\ngonum.org/v1/gonum/graph/iterator.(*Nodes).Next(0xc00157bc50)\n\tgonum.org/v1/gonum/graph/iterator/nodes_map.go:40 +0x36\ngithub.com/grafana/grafana/pkg/expr.buildNodeRegistry(0xc0009a0d60)\n\tgithub.com/grafana/grafana/pkg/expr/graph.go:115 +0x149\ngithub.com/grafana/grafana/pkg/expr.(*Service).buildDependencyGraph(0xc00191f288?, 0x26e34b7?)\n\tgithub.com/grafana/grafana/pkg/expr/graph.go:85 +0x28\ngithub.com/grafana/grafana/pkg/expr.(*Service).buildPipeline(0xc08999d8dd051dd9?, 0x269979e3b?)\n\tgithub.com/grafana/grafana/pkg/expr/graph.go:65 +0x19\ngithub.com/grafana/grafana/pkg/expr.(*Service).BuildPipeline(...)\n\tgithub.com/grafana/grafana/pkg/expr/service.go:61\ngithub.com/grafana/grafana/pkg/expr.(*Service).TransformData(0xc0019e9f80, {0x108d3f0, 0xc00157c360}, 0xc00157b140)\n\tgithub.com/grafana/grafana/pkg/expr/transform.go:134 +0x114\ngithub.com/grafana/grafana/pkg/services/ngalert/eval.executeQueriesAndExpressions({0x1, 0x1, {0x82c372fd8, 0xc0019f26a0}, {0x108d3f0, 0xc00157c360}}, {0xc000fc7200?, 0x4abe160?, 0x5348e0?}, {0xc08999d8dd051dd9, ...}, ...)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/eval/eval.go:274 +0x1ce\ngithub.com/grafana/grafana/pkg/services/ngalert/eval.executeCondition({0x1, 0x1, {0x82c372fd8, 0xc0019f26a0}, {0x108d3f0, 0xc00157c360}}, 0xc00191fc18, {0xc08999d8dd051dd9, 0x269979e3b, 0x4abe160}, ...)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/eval/eval.go:173 +0xed\ngithub.com/grafana/grafana/pkg/services/ngalert/eval.(*Evaluator).ConditionEval(0xc0009324d8, 0xc00191fc18, {0x29ef6f2?, 0xc0?, 0x4abe160?}, 0xd0?)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/eval/eval.go:510 +0x16f\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine.func2(0xc0005010e0, 0xacd84f?, 0xc0009a0940)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:460 +0x30a\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine.func4.2(0x38?)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:556 +0x1b7\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine.func3(0xc0011cec00)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:517 +0x5c\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine.func4(0x0?, 0xc001510d00?, {0xc001174d01?, {0xc0012f3380?, 0xc001510d00?}}, 0xc0009a0940, 0xc000a39eb0, 0xc00099f368, 0xc0011cebc0, {0x109b0b0, ...}, ...)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:546 +0x194\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine(0xc000932480, {0x108d380, 0xc0011ce880}, {0x265f220?, {0xc0012f3380?, 0xc001174f88?}}, 0xc001775f80)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:561 +0x6b5\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).schedulePeriodic.func1()\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:367 +0x38\ngolang.org/x/sync/errgroup.(*Group).Go.func1()\n\tgolang.org/x/sync/errgroup/errgroup.go:57 +0x67\ncreated by golang.org/x/sync/errgroup.(*Group).Go\n\tgolang.org/x/sync/errgroup/errgroup.go:54 +0x8d\n"
```

Grafana version: 8.3.6
Data source type & version: InfluxDB 1.8.10
OS Grafana is installed on: FreeBSD 13

Would be interesting to see if that also happens to linux users.

**Comment by armandgrillet** [comment]
https://github.com/grafana/grafana/pull/46997 fixes the issue in 8.4, 8.5, and forward.

**Comment by drTr0jan** [comment]
@jan666, is there the error while build on Go 1.17?
Try to test [a port patch](https://bz-attachments.freebsd.org/attachment.cgi?id=233186) from [bug #263251](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=263251).

**Comment by jan666** [comment]
I updated to 8.4.6 (using go 1.18 again, using the patch from https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=263251), same issue:

`logger=ngalert t=2022-04-14T07:27:28.62+0200 lvl=eror msg="alert rule panic" error="runtime error: invalid memory address or nil pointer dereference" stack="goroutine 638 [running]:\nruntime/debug.Stack()\n\truntime/debug/stack.go:24 +0x65\ngithub.com/grafana/grafana/pkg/services/ngalert/eval.executeQueriesAndExpressions.func1()\n\tgithub.com/grafana/grafana/pkg/services/ngalert/eval/eval.go:334 +0x5d\npanic({0x5dade0, 0x4987980})\n\truntime/panic.go:838 +0x207\nreflect.mapiterinit(0x98b460?, 0xc00136d140?, 0xc00136d140?)\n\truntime/map.go:1373 +0x19\ngonum.org/v1/gonum/graph/iterator.(*mapIter).next(0xc00133c270)\n\tgonum.org/v1/gonum/graph/iterator/map.go:89 +0x5e\ngonum.org/v1/gonum/graph/iterator.(*Nodes).Next(0xc001cfa450)\n\tgonum.org/v1/gonum/graph/iterator/nodes_map.go:40 +0x36\ngithub.com/grafana/grafana/pkg/expr.buildNodeRegistry(0xc000e4ca80)\n\tgithub.com/grafana/grafana/pkg/expr/graph.go:115 +0x149\ngithub.com/grafana/grafana/pkg/expr.(*Service).buildDependencyGraph(0xc00136d208?, 0x27b0117?)\n\tgithub.com/grafana/grafana/pkg/expr/graph.go:85 +0x28\ngithub.com/grafana/grafana/pkg/expr.(*Service).buildPipeline(0x50e7fd57f?, 0x4c376a0?)\n\tgithub.com/grafana/grafana/pkg/expr/graph.go:65 +0x19\ngithub.com/grafana/grafana/pkg/expr.(*Service).BuildPipeline(...)\n\tgithub.com/grafana/grafana/pkg/expr/service.go:61\ngithub.com/grafana/grafana/pkg/expr.(*Service).TransformData(0xc0000f9d70, {0x10dcf18, 0xc001b52240}, 0xc00199b080)\n\tgithub.com/grafana/grafana/pkg/expr/transform.go:79 +0x114\ngithub.com/grafana/grafana/pkg/services/ngalert/eval.executeQueriesAndExpressions({0x1, 0x1, {0x10e5bb0, 0xc0010d3b00}, {0x10dcf18, 0xc001b52240}}, {0xc00156e300?, 0xc001266640?, 0x0?}, {0xc08e0a1024361d32, ...}, ...)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/eval/eval.go:349 +0x225\ngithub.com/grafana/grafana/pkg/services/ngalert/eval.executeCondition({0x1, 0x1, {0x10e5bb0, 0xc0010d3b00}, {0x10dcf18, 0xc001b52240}}, 0xc00136db88, {0xc08e0a1024361d32, 0x50e7fd57f, 0x4c376a0}, ...)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/eval/eval.go:248 +0x125\ngithub.com/grafana/grafana/pkg/services/ngalert/eval.(*Evaluator).ConditionEval(0xc001247040, 0xc00136db88, {0x6?, 0x6?, 0x4c376a0?}, 0xc001aff200?)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/eval/eval.go:585 +0x1ae\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine.func4({0x10dcea8, 0xc000f124c0}, 0xc00157eff0, 0x10d0d40?, 0xc000e4c740)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:557 +0x2fe\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine.func7.2(0x40?)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:636 +0x1e6\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine.func5(0xc000ed4b00)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:581 +0x5c\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine.func7(0x498c1c0?, 0x1?, {0x0?, {0xc0012f1740?, 0x0?}}, 0xc000e4c740, 0xc0012a9e68, 0xc0000afba0, 0xc000f131c0, {0x10dcea8, ...}, ...)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:626 +0x1b6\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine(0xc001238ea0, {0x10dcea8, 0xc000f124c0}, {0x3f31ce0?, {0xc0012f1740?, 0xc00087cea0?}}, 0xc00087cf60, 0xc00087cfc0)\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:641 +0x87b\ngithub.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).schedulePeriodic.func1()\n\tgithub.com/grafana/grafana/pkg/services/ngalert/schedule/schedule.go:415 +0x3c\ngolang.org/x/sync/errgroup.(*Group).Go.func1()\n\tgolang.org/x/sync/errgroup/errgroup.go:57 +0x67\ncreated by golang.org/x/sync/errgroup.(*Group).Go\n\tgolang.org/x/sync/errgroup/errgroup.go:54 +0x8d\n"`

**Comment by drTr0jan** [comment]
@jan666, of course. The 8.4.6 port version doesn't contain #46997.
Try to use [8.3.6_2 port version](https://bz-attachments.freebsd.org/attachment.cgi?id=233186) with Go 1.17.

**Comment by jan666** [comment]
@drTr0jan yea using go117 works (was my workaround before - manually install go117 again and rebuild grafana). 

I'm using 8.4.6 built with go117 now. See FreeBSD Bugzilla

**Comment by jan666** [comment]
This is fixed in 8.4.7. No more crashes on startup when built with go 1.18

FreeBSD port patch: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=263438
